### PR TITLE
Support for experimental BuildKit

### PR DESCRIPTION
**Pull request by tiborvass**
This PR makes `docker build` use the experimental support in moby/moby#37151 when the `DOCKER_BUILDKIT` environment variable is set.

This is still a work in progress.

To test it, build the docker CLI from this PR, and use it with the engine built from moby/moby#37151.
Mainly looking for UX feedback and bad bugs.

The output of the builder is changing (using that of buildkit for interactive mode). Non-interactive output is currently JSON stream for now but it will likely evolve to a more human-readable output (the challenge being handling output of parallel RUN commands).

Known issues:
- no support for `--iidfile` (coming soon)
- no support for docker content trust (out of scope for this PR)

Ping @tonistiigi @thaJeztah @vdemeester @garethr

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -52,7 +52,7 @@ watch: ## monitor file changes and run go test
 
 vendor: vendor.conf ## check that vendor matches vendor.conf
 	rm -rf vendor
-	bash -c 'vndr |& grep -v -i clone'
+	bash -c 'vndr -verbose'
```
Comment (simonferquel) [inline Makefile]: Is this change related to the PR?

File: cli/command/image/build/context.go [hunk 2]
```diff
@@ -81,59 +81,88 @@ func ValidateContextDirectory(srcPath string, excludes []string) error {
...
+// DetectArchiveReader detects whether the input stream is an archive or a
+// Dockerfile and returns a buffered version of input, safe to consume in lieu
+// of input. If an archive is detected, isArchive is set to true, and to false
+// otherwise, in which case it is safe to assume input represents the contents
+// of a Dockerfile.
+//
+// If an archive is detected in the input stream, rc stream is non-nil and ready
+// to be consumed in lieu of the input stream, and dockerfileDir is empty.
+//
+// If a Dockerfile is detected, rc is set to nil and dockerfileDir is set to a
```
Comment (simonferquel) [inline cli/command/image/build/context.go]: Comment does not seem in sync

File: cli/command/image/build/context.go [hunk 3]
```diff
@@ -81,59 +81,88 @@ func ValidateContextDirectory(srcPath string, excludes []string) error {
...
-		return nil, "", errors.New("build context is not an archive")
-	}
 
-	// Input should be read as a Dockerfile.
-	tmpDir, err := ioutil.TempDir("", "docker-build-context-")
+// WriteTempDockerfile writes a Dockerfile stream to a temporary file with a
+// name specified by DefaultDockerfileName and returns the path to the
+// temporary directory containing the Dockerfile.
+func WriteTempDockerfile(rc io.ReadCloser) (dockerfileDir string, err error) {
+	dockerfileDir, err = ioutil.TempDir("", "docker-build-tempdockerfile-")
```
Comment (simonferquel) [inline cli/command/image/build/context.go]: error management in this function seem more complicated than necessary. As the err output parameter is named, why not use a single defer block to centralize error reporting ?

**Comment by tiborvass** [comment]
@simonferquel updated. Appreciate your review!

File: cli/command/image/build/context.go [hunk 4]
```diff
@@ -81,59 +81,81 @@ func ValidateContextDirectory(srcPath string, excludes []string) error {
...
-	f, err := os.Create(filepath.Join(tmpDir, DefaultDockerfileName))
+	f, err := os.Create(filepath.Join(dockerfileDir, DefaultDockerfileName))
 	if err != nil {
-		return nil, "", err
+		return "", err
 	}
-	_, err = io.Copy(f, buf)
+	_, err = io.Copy(f, rc)
 	if err != nil {
 		f.Close()
```
Comment (simonferquel) [inline cli/command/image/build/context.go]:
```
defer func (){
  er := f.Close ()
  if err == nil {
    err = er
  }
}
```
Not sure about it, but I dislike having duplicated Close call. Seem error prone if/when new code paths are introduced in the future.

File: cli/command/image/build.go [hunk 5]
```diff
@@ -149,6 +152,9 @@ func NewBuildCommand(dockerCli command.Cli) *cobra.Command {
 	flags.SetAnnotation("stream", "experimental", nil)
 	flags.SetAnnotation("stream", "version", []string{"1.31"})
 
+	flags.BoolVar(&options.noConsole, "no-console", false, "Show non-console output")
```
Comment (AkihiroSuda) [inline cli/command/image/build.go]: Maybe better to add "BuildKit mode only"?

File: cli/command/image/build.go [hunk 6]
```diff
@@ -170,6 +176,10 @@ func (out *lastProgressOutput) WriteProgress(prog progress.Progress) error {
 
 // nolint: gocyclo
 func runBuild(dockerCli command.Cli, options buildOptions) error {
+	if os.Getenv("DOCKER_BUILDKIT") != "" {
+		return runBuildBuildKit(dockerCli, options)
```
Comment (AkihiroSuda) [inline cli/command/image/build.go]: Any reason not to use CLI flag?

File: cli/command/image/build.go (hunk 5 above)
Comment (tiborvass) [reply cli/command/image/build.go]: Sure. BTW, the main reason we have this flag, is because the interactive output of buildkit doesn't show outputs of RUN commands. If it did, we might not need this flag. Which makes me wonder whether we could remove this flag in the future as it graduates from experimental. cc @tonistiigi @vdemeester

File: cli/command/image/build.go (hunk 6 above)
Comment (tiborvass) [reply cli/command/image/build.go]: @AkihiroSuda because in the future I'd like to remove it.

**Comment by tiborvass** [comment]
Ping @simonferquel if you have any further comments.

**Comment by simonferquel** [comment]
Lgtm :)

File: cli/command/image/build/context.go [hunk 7]
```diff
@@ -81,59 +81,76 @@ func ValidateContextDirectory(srcPath string, excludes []string) error {
...
+	f, err := os.Create(filepath.Join(dockerfileDir, DefaultDockerfileName))
 	if err != nil {
-		return nil, "", errors.Errorf("unable to create temporary context directory: %v", err)
+		return "", err
+	}
+	defer f.Close()
+	if _, err := io.Copy(f, rc); err != nil {
+		return "", err
 	}
+	return dockerfileDir, rc.Close()
```
Comment (ijc) [inline cli/command/image/build/context.go]: If `rc.Close()` returns an error then the callers will leak `tmpDir`, since they don't clean it up in the error case (and it would be a little odd to expect them to do so).

The dir is also leaked on errors in in the `os.Create` above too.

File: cli/command/image/build_buildkit.go [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+		rc, isArchive, err := build.DetectArchiveReader(os.Stdin)
+		if err != nil {
+			return err
+		}
+		if isArchive {
+			body = rc
+			remote = uploadRequestRemote
+		} else {
+			if options.dockerfileName != "" {
+				// This is new behavior, but @tiborvass believes it's correct.
```
Comment (ijc) [inline cli/command/image/build_buildkit.go]: This looks like a discussion for the commit message rather than a code comment.

File: cli/command/image/build/context.go (hunk 7 above)
Comment (tiborvass) [reply cli/command/image/build/context.go]: @ijc Addressed.

File: cli/command/image/build_buildkit.go (hunk 8 above)
Comment (tiborvass) [reply cli/command/image/build_buildkit.go]: @ijc addressed in commit message.

File: cli/command/image/build_buildkit.go [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+			},
+		}))
+	}
+
+	// statusContext, cancelStatus := context.WithCancel(ctx)
+	// defer cancelStatus()
+
+	// if span := opentracing.SpanFromContext(ctx); span != nil {
+	// 	statusContext = opentracing.ContextWithSpan(statusContext, span)
+	// }
```
Comment (thaJeztah) [inline cli/command/image/build_buildkit.go]: Should this be removed before merging?

File: cli/command/image/build_buildkit.go [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+	displayStatus := func(displayCh chan *client.SolveStatus) {
+		var c console.Console
+		out := os.Stderr
+		// TODO: Handle interactive output in non-interactive environment.
+		consoleOpt := options.console.Value()
+		if cons, err := console.ConsoleFromFile(out); err == nil && (consoleOpt == nil || *consoleOpt) {
+			c = cons
+		}
+		// not using shared context to not disrupt display but let is finish reporting errors
```
Comment (thaJeztah) [inline cli/command/image/build_buildkit.go]: typo: s/let is/let it/

File: cli/command/image/build_buildkit.go [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+	err = jsonmessage.DisplayJSONMessagesStream(response.Body, os.Stdout, dockerCli.Out().FD(), dockerCli.Out().IsTerminal(), t.write)
+	if err != nil {
+		if jerr, ok := err.(*jsonmessage.JSONError); ok {
+			// If no error code is set, default to 1
+			if jerr.Code == 0 {
+				jerr.Code = 1
+			}
+			// if options.quiet {
+			// 	fmt.Fprintf(dockerCli.Err(), "%s%s", progBuff, buildBuff)
+			// }
```
Comment (thaJeztah) [inline cli/command/image/build_buildkit.go]: should this be removed before merging? (or a TODO added)

File: cli/command/formatter/disk_usage.go [hunk 12]
```diff
@@ -377,17 +396,30 @@ func (c *diskUsageBuilderContext) Type() string {
...
+	return fmt.Sprintf("%d", numActive)
 }
 
 func (c *diskUsageBuilderContext) Size() string {
 	return units.HumanSize(float64(c.builderSize))
 }
 
 func (c *diskUsageBuilderContext) Reclaimable() string {
-	return c.Size()
+	inUseBytes := int64(0)
```
Comment (thaJeztah) [inline cli/command/formatter/disk_usage.go]: `var inUseBytes int64` ?

File: cli/command/image/build/context.go (hunk 3 above)
Comment (thaJeztah) [reply cli/command/image/build/context.go]: nit: consider renaming the `err` output var, to prevent it from being confused with local vars (we ran into a couple of cases where this was overlooked in code changes)

File: cli/command/image/build_buildkit.go [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+//nolint: gocyclo
+func runBuildBuildKit(dockerCli command.Cli, options buildOptions) error {
+	ctx := appcontext.Context()
+
+	s, err := trySession(dockerCli, options.context)
+	if err != nil {
+		return err
+	}
+	if s == nil {
+		return errors.Errorf("buildkit not supported by daemon")
```
Comment (thaJeztah) [inline cli/command/image/build_buildkit.go]: can we make `trysession` return this error?

File: cli/command/image/build_buildkit.go [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+	}
+
+	buildID := stringid.GenerateRandomID()
```
Comment (thaJeztah) [inline cli/command/image/build_buildkit.go]: looks like this isn't used until quite some lines below; could you move this closer to where it's used?

File: cli/command/image/build_buildkit.go [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+			// If no error code is set, default to 1
+			if jerr.Code == 0 {
+				jerr.Code = 1
+			}
+			// if options.quiet {
+			// 	fmt.Fprintf(dockerCli.Err(), "%s%s", progBuff, buildBuff)
+			// }
+			return cli.StatusError{Status: jerr.Message, StatusCode: jerr.Code}
+		}
+		return err
```
Comment (thaJeztah) [inline cli/command/image/build_buildkit.go]: this is redundant if you change the return below to `return err`.

perhaps changing the `if err != nil` to `if err == nil` above (and return early), could make the logic a bit less complicated

File: cli/command/image/build_buildkit.go [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+	}
+
+	return nil
+}
+
+func resetUIDAndGID(s *fsutil.Stat) bool {
+	s.Uid = uint32(0)
+	s.Gid = uint32(0)
```
Comment (thaJeztah) [inline cli/command/image/build_buildkit.go]: is the type conversion needed here? (just wondering)

File: cli/command/image/build_buildkit.go [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+	return true
```
Comment (thaJeztah) [inline cli/command/image/build_buildkit.go]: was wondering why the boolean return was here, but guess this is to satisfy an interface? (reading from my phone so couldn't check easily)

File: cli/command/image/build_buildkit.go [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+
+	if msg.ID != "moby.buildkit.trace" {
+		return
+	}
+
+	var dt []byte
+	if err := json.Unmarshal(*msg.Aux, &dt); err != nil {
+		return
+	}
+	if err := (&resp).Unmarshal(dt); err != nil {
```
Comment (thaJeztah) [inline cli/command/image/build_buildkit.go]: it's ok to hide them actual errors? May be good to document that (as a comment here and/or godoc on the function)

**Comment by andrewhsu** [comment]
Stuff still works with 5a103e1 with my manual build. I also ran a docker-ce integration test suite and stuff is [green](https://jenkins.dockerproject.org/job/docker-ce/372/console).

**Comment by tiborvass** [comment]
@thaJeztah updated

File: cli/command/image/build/context.go [hunk 19]
```diff
@@ -81,59 +81,82 @@ func ValidateContextDirectory(srcPath string, excludes []string) error {
...
+	return ioutils.NewReadCloserWrapper(buf, func() error { return input.Close() }), IsArchive(magic), nil
+}
 
-	if dockerfileName == "-" {
-		return nil, "", errors.New("build context is not an archive")
+// WriteTempDockerfile writes a Dockerfile stream to a temporary file with a
+// name specified by DefaultDockerfileName and returns the path to the
+// temporary directory containing the Dockerfile.
+func WriteTempDockerfile(rc io.ReadCloser) (dockerfileDir string, err error) {
+	// err is a named return value, due to the defer call below.
```
Comment (thaJeztah) [inline cli/command/image/build/context.go]: was thinking naming it `retErr` (eg), but just a nit; no need to change

**Comment by tonistiigi** [comment]
LGTM
